### PR TITLE
Copy `bin/*` and `lib/*.dylib` files to `stage0-sysroot`

### LINKED ISSUE #101691: Copy stage0 binaries into stage0-sysroot
Currently, the beta binaries are only present in `build/stage0/bin`; `build/stage0-sysroot` has *only* the standard library, not any binaries. That causes two issues:
1. `rustup toolchain link` doesn't work with stage0-sysroot, meaning there's no way to change the standard library and use those changes at runtime without doing a full stage 1 build
2. `x dist --stage 0` doesn't work. This is very rare to want to do, but can be useful when modifying the `dist` process itself, to avoid long compile times.
 ```
thread 'main' panicked at 'could not read dir "/home/dawn/projects/rust/build/x86_64-unknown-linux-gnu/stage0-sysroot/bin": Os { code: 2, kind: NotFound, message: "No such file or directory" }', lib.rs:1573:25
```
 
Copying the binaries into stage0-sysroot is an easy way to fix both problems at once.

Mentoring instructions: Add a `builder.copy` call in `impl Step for Sysroot` https://github.com/rust-lang/rust/blob/6f0c4a6c5c36f1f8f433a12e10a29918f3d40a31/src/bootstrap/compile.rs#L1146 which copies the whole `bin` directory from `build/stage0`.

### PR DESCRIPTION
**Pull request by KittyBorgX**
Fixes #101691

### CODE REVIEW

**Comment by jyn514** [comment]
I don't have time for reviews right now.

r? @ozkanonur

**Comment by KittyBorgX** [comment]
@ozkanonur I've fixed the breaking changes. Do I add a new commit or ammend it to the existing one?

**Comment by ozkanonur** [comment]
> @ozkanonur I've fixed the breaking changes. Do I add a new commit or ammend it to the existing one?

Ammend it please :)

**Comment by ozkanonur** [comment]
ref from the issue:
> 2. `x dist --stage 0` doesn't work:
> 
> ```
> thread 'main' panicked at 'could not read dir "/home/dawn/projects/rust/build/x86_64-unknown-linux-gnu/stage0-sysroot/bin": Os { code: 2, kind: NotFound, message: "No such file or directory" }', lib.rs:1573:25
> ```


Still doesn't work

```sh
 remotes/KittyBorgX/master $ ./x dist --stage 0

thread 'main' panicked at 'could not read dir "/home/nimda/devspace/personal/rust2/build/x86_64-unknown-linux-gnu/stage0-sysroot/lib/rustlib/x86_64-unknown-linux-gnu/codegen-backends": Os { code: 2, kind: NotFound, message: "No such file or directory" }', lib.rs:1522:25
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

**Comment by KittyBorgX** [comment]
I see the issue. The original pr (#101711) broke when `rustc` and `cargo` was set which works fine with my PR. 
Here's a screenshot: 
image

As to `codegen-backends` which failed in the above comment: 
```
 remotes/KittyBorgX/master $ ./x dist --stage 0
thread 'main' panicked at 'could not read dir "/home/nimda/devspace/personal/rust2/build/x86_64-unknown-linux-gnu/stage0-sysroot/lib/rustlib/x86_64-unknown-linux-gnu/codegen-backends": Os { code: 2, kind: NotFound, message: "No such file or directory" }', lib.rs:1522:25
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

I followed : https://rust-lang.zulipchat.com/#narrow/stream/326414-t-infra.2Fbootstrap/topic/Copying.20binaries.20to.20stage0-sysroot/near/326813495

and didn't copy it. I will work on the same and update this pr. 
Sidenote : On like 463, in my current pr, it is `x.starts_with(sysroot_lib_dir.join("rustlib").into_boxed_path())` which passes the ci and passes the test locally but doesn't do the actual job that is intended and fails like how you mentioned in the second half of the above comment. But having `!x.starts_with(sysroot_lib_dir.join("rustlib").into_boxed_path())` does the job of copying correctly but fails the CI and doesn't pass the tests. This has hit me with a roadblock which forces me to use another option and look into it deeply. Can I take some more time to work on this issue, if that's alright?

**Comment by ozkanonur** [comment]
> I see the issue. The original pr (#101711) broke when `rustc` and `cargo` was set which works fine with my PR. Here's a screenshot: image

This doesn't work on `x86_64`. 
```sh
  ~  $ rustc +issue-101691 --v
/home/nimda/.rustup/toolchains/issue-101691/bin/rustc: error while loading shared libraries: librustc_driver-c78d16bcb49dc5d0.so: cannot open shared object file: No such file or directory
```

> Can I take some more time to work on this issue, if that's alright?

Please take your time

**Comment by KittyBorgX** [comment]
@ozkanonur Can you help me? I've found a fix which works and passes the tests but I pulled the upstream changes and git(hub) added a commit. Is there any way I can delete the commit? Or should I close this PR and open a new one?

**Comment by KittyBorgX** [comment]
Thank you so much! I've done the necessary changes and pushed it.

**Comment by ozkanonur** [comment]
> Thank you so much! I've made the necessary changes and pushed it.

Great! I will review it later today or tomorrow.

**Comment by KittyBorgX** [comment]
Thanks! 
Just a few points: 
- I didn't run tidy and fmt which caused the ci to break :( 
- I forgot to take care of the `codegen-backends` and `./x dist --stage 0` might fail. I have to code ready for it, I'll format it, tidy it and push it asap after the ci is done building.

**Comment by ozkanonur** [comment]
:(

```sh
~/devspace/personal/rust-dist/compiler  $ ../../rust/x dist --stage 0

thread 'main' panicked at 'generated a symlink in a tarball: /home/nimda/devspace/personal/rust-dist/compiler/build/tmp/tarball/rustc/x86_64-unknown-linux-gnu/rustc-1.69.0-dev-x86_64-unknown-linux-gnu/rustc/bin/rustfmt', tarball.rs:330:21
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
        finished in 34.565 seconds
Build completed unsuccessfully in 0:04:02
```

File: src/bootstrap/compile.rs [hunk 1]
```diff
@@ -445,6 +445,48 @@ impl Step for StdLink {
...
+            let sysroot = builder.out.join(&compiler.host.triple).join("stage0-sysroot");
+
+            let host_lib_dir = builder.initial_rustc.ancestors().nth(2).unwrap().join("lib");
+            let host_bin_dir = builder.out.join(&builder.initial_rustc.parent().unwrap());
+            let host_codegen_backends = builder
+                .initial_rustc
+                .ancestors()
+                .nth(2)
+                .unwrap()
+                .join("lib/rustlib")
```
Comment (ozkanonur) [inline src/bootstrap/compile.rs:460]:
```suggestion
            let host_codegen_backends = host_lib_dir
                .join("rustlib")
```

**Review by onur-ozkan — CHANGES_REQUESTED** [review]

File: src/bootstrap/compile.rs [hunk 2]
```diff
@@ -445,6 +445,48 @@ impl Step for StdLink {
...
+                    {
+                        builder.copy(&path, &sysroot_lib_dir.join(path.file_name().unwrap()));
+                    }
+                }
+            }
+            // copy bin files from `builder.initial_rustc/./` to `stage0-sysroot/bin`
+            builder.cp_r(&host_bin_dir, &sysroot_bin_dir);
+
+            // copy bin files from `builder.initial_rustc/./` to `stage0-sysroot/bin`
+            builder.cp_r(&host_codegen_backends, &sysroot_codegen_backends);
```
[the remark this answers is not in the document]
Comment (KittyBorgX) [reply src/bootstrap/compile.rs]: I copy pasted to comment and forgot to change it!
Fixed it now.

File: src/bootstrap/compile.rs (hunk 1 above)
Comment (KittyBorgX) [reply src/bootstrap/compile.rs]: Done!

**Comment by KittyBorgX** [comment]
> :(
> 
> ```shell
> ~/devspace/personal/rust-dist/compiler  $ ../../rust/x dist --stage 0
> 
> thread 'main' panicked at 'generated a symlink in a tarball: /home/nimda/devspace/personal/rust-dist/compiler/build/tmp/tarball/rustc/x86_64-unknown-linux-gnu/rustc-1.69.0-dev-x86_64-unknown-linux-gnu/rustc/bin/rustfmt', tarball.rs:330:21
> note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
>         finished in 34.565 seconds
> Build completed unsuccessfully in 0:04:02
> ```

a- 
im lost on how to debug this but I will get this done in 12 hours ( 8 hours of which is my sleep :) )

**Comment by KittyBorgX** [comment]
@ozkanonur Im really sorry if this is taking longer than expected, this is my first contribution and im half new to rust and the rust repo :(

**Comment by ozkanonur** [comment]
> @ozkanonur Im really sorry if this is taking longer than expected, this is my first contribution and im half new to rust and the rust repo :(

It's all fine, don't worry about it :)

You can do the tests and once you see they all working fine, push the changes and wait for CI pipelines to be passed. If everything goes well, ping me here saying "you can do the review" so I know when to review this PR.

You can do the tests with the following instructions:
```sh
## 1st test ##

# make sure this doesn't panic
x dist --stage 0

## 2nd test ##

# build stage 0
x build --stage 0

# then link the stage0-sysroot
rustup toolchain link {toolchain_name} {path_to_stage0-sysroot}

# check if rustc works properly with linked toolchain
rustc +{toolchain_name} --version

# sanity check for rustc, basically create simple rust project and run it with linked toolchain
cargo new hello
cd hello
cargo +{toolchain_name} run
```

**Comment by KittyBorgX** [comment]
@ozkanonur 
Here's the update: 
I got `x dist --stage 0` to finally build successfully. However as `@jyn514` mentioned, there's a bug in `dist` which causes the submodules to not get checked out automatically. I'm filing an issue for that today. Do I wait for that issue to get resolved or shall I push my changes here? I got it to work by adding some more code and running `git submodule update --init`

PS : Here's the working screenshot: 
image

**Comment by KittyBorgX** [comment]
@rustbot ready
@ozkanonur You can review now :) I'm done with all my changes.

File: src/bootstrap/compile.rs [hunk 3]
```diff
@@ -445,6 +445,70 @@ impl Step for StdLink {
         let libdir = builder.sysroot_libdir(target_compiler, target);
         let hostdir = builder.sysroot_libdir(target_compiler, compiler.host);
         add_to_sysroot(builder, &libdir, &hostdir, &libstd_stamp(builder, compiler, target));
+        if compiler.stage == 0 {
```
Comment (ozkanonur) [inline src/bootstrap/compile.rs:448]:
```suggestion

        if compiler.stage == 0 {
```

File: src/bootstrap/compile.rs [hunk 4]
```diff
@@ -445,6 +445,70 @@ impl Step for StdLink {
...
+                            );
+                        }
+                    } else {
+                        builder.copy(
+                            &file.path().into_boxed_path(),
+                            &sysroot_bin_dir.join(file.file_name()),
+                        );
+                    }
+                }
+            }
```
Comment (ozkanonur) [inline src/bootstrap/compile.rs:491]:
```suggestion
            }

```

**Review by onur-ozkan — APPROVED** [review]
this looks good, can you add this empty lines as well?

**Comment by KittyBorgX** [comment]
Done 👍

**Comment by KittyBorgX** [comment]
@ozkanonur and @jyn514  Tysm for helping me throught the PR! I couldn't have done it without your help :D

**Comment by ozkanonur** [comment]
congratz your first PR :)

**Comment by albertlarsan68** [comment]
@rust-timer build 89394208c043edd910fa6c950968ca75fcf349c0

**Comment by albertlarsan68** [comment]
@KittyBorgX ^
This PR does break rustc benches, but not the bootstrap timings, it has them.

**Comment by KittyBorgX** [comment]
ah alright. 
@albertlarsan68 thank you so much for running it :)

**Comment by Kobzol** [comment]
It's the same thing AFAIK, bootstrap timings = building rustc. The timings there are from a previous run, the "B column" is missing.

**Pull request merged**